libs/sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java

```diff
@@ -511,6 +512,49 @@ public void test_special_char_data_type() {
         assertThat(cast.getType().name()).isEqualTo("\"char\"");
     }
 
+    @Test
+    public void test_dollar_quoted_strings_with_valid_tags() {
+        assertThat(((StringLiteral) SqlParser.createExpression("$$$$")).getValue()).isEqualTo("");
+        assertThat(((StringLiteral) SqlParser.createExpression("$a$ a$a $a$")).getValue()).isEqualTo(" a$a ");
+        assertThat(((StringLiteral) SqlParser.createExpression("$a$$a$")).getValue()).isEqualTo("");
+        assertThat(((StringLiteral) SqlParser.createExpression("$_$$_$")).getValue()).isEqualTo("");
+        assertThat(((StringLiteral) SqlParser.createExpression("$_$string, content,\n!@#$_$")).getValue()).isEqualTo("string, content,\n!@#");
+    }
+
+    @Test
+    public void test_comma_separated_dollar_quoted_strings() {
+        // tests that '$$a$$, $$b$$' is not interpreted as 'a$$, $$b'
+        var arrayLiteral = (ArrayLiteral) SqlParser.createExpression("[$$a$$, $$b,$$, $a$abc$a$]");
+        var actual = arrayLiteral.values().stream().map(expr -> (StringLiteral) expr).map(StringLiteral::getValue).toList();
+        assertThat(actual).isEqualTo(List.of("a", "b,", "abc"));
+    }
+
+    @Test
+    public void test_dollar_quoted_positional_parameters() {
+        assertThat(((StringLiteral) SqlParser.createExpression("$$$1$$")).getValue()).isEqualTo("$1");
+        assertThat(((StringLiteral) SqlParser.createExpression("$$$1$2$3$$")).getValue()).isEqualTo("$1$2$3");
+    }
+
+    @Test
+    public void test_dollar_quoted_strings_with_invalid_tags() {
+        assertThatThrownBy(
+            () -> SqlParser.createExpression("$a$a$a$a$")) // tokenize as $a$ | a | $a$ then the next 'a' is mismatched
+            .isInstanceOf(ParsingException.class)
+            .hasMessageStartingWith("line 1:8: mismatched input 'a' expecting");
+        assertThatThrownBy(
+            () -> SqlParser.createExpression("$$$$$$"))
+            .isInstanceOf(ParsingException.class)
+            .hasMessageStartingWith("line 1:5: extraneous input '$$' expecting");
+        assertThatThrownBy(
+            () -> SqlParser.createExpression("$1$$1$")) // $1$ is an invalid tag
+            .isInstanceOf(ParsingException.class)
+            .hasMessageStartingWith("line 1:3: mismatched input '$$' expecting");
+        assertThatThrownBy(
+            () -> SqlParser.createExpression("$a$$b$"))
+            .isInstanceOf(ParsingException.class)
+            .hasMessageStartingWith("line 1:7: extraneous input '<EOF>' expecting {DOLLAR_QUOTED_STRING_BODY, END_DOLLAR_QUOTED_STRING}");
```

Comment:
Imho, this is not nice for the user, can we have the old behaviour where we list the exact tokens and not the lexer names?

Update, now I get why the change in my other comments.

Reply:
To my understanding, the behaviour did not change. Before, `mismatched input '<EOF>' expecting {'(', '[', ...` message looked like it contained the exact tokens because '(`, `[`, etc were implicitly defined.

Reply:
Can you take a look here: https://www.antlr2.org/doc/err.html#Modifying%20Default%20Error%20Messages%20With%20Paraphrases ?

Reply:
hm, this is for antlr2, but maybe something like that exists for antlr4

Reply:
Thank you, I will have a look at it now. Just to clarify, instead of `expecting {DOLLAR_QUOTED_STRING_BODY` would you expect to see `expecting { ~ '$'+ }` ?

Reply:
The best I could find is this: https://github.com/crate/crate/commit/9ed5b86cd9acec8fd301e998b050f8c523863fe2
I've also replaced some left over literals in the parser grammar with token names from lexer.

Reply:
> Thank you, I will have a look at it now. Just to clarify, instead of `expecting {DOLLAR_QUOTED_STRING_BODY` would you expect to see `expecting { ~ '$'+ }` ?

Something like that yes, or check what I've used in my commit ^^.

Reply:
Thanks, it looks good. I pushed it